editor/src/components/canvas/controls/select-mode/drag-outline-control.tsx

```diff
@@ -1,50 +1,96 @@
 import React from 'react'
+import {
+  boundingRectangleArray,
+  CanvasRectangle,
+  offsetRect,
+} from '../../../../core/shared/math-utils'
 import { ElementPath } from '../../../../core/shared/project-file-types'
+import { assertNever } from '../../../../core/shared/utils'
 import { useColorTheme } from '../../../../uuiui'
 import { useEditorState } from '../../../editor/store/store-hook'
 import { controlForStrategyMemoized } from '../../canvas-strategies/canvas-strategy-types'
 import { getMultiselectBounds } from '../../canvas-strategies/strategies/shared-move-strategies-helpers'
 import { CanvasOffsetWrapper } from '../canvas-offset-wrapper'
 
-interface DragOutlineControlProps {
+export const DragOutlineControlTestId = 'drag-outline-control'
+
+interface DragTargetsElementPaths {
+  type: 'element-paths'
   targets: Array<ElementPath>
 }
 
-export const DragOutlineControl = controlForStrategyMemoized(
-  ({ targets }: DragOutlineControlProps) => {
-    const scale = useEditorState((store) => store.editor.canvas.scale, 'OutlineControl scale')
-    const frame = useEditorState((store) => {
-      return getMultiselectBounds(store.strategyState.startingMetadata, targets)
-    }, 'GhostOutline frame')
-    const dragVector = useEditorState((store) => {
-      if (store.editor.canvas.interactionSession?.interactionData.type === 'DRAG') {
-        return store.editor.canvas.interactionSession.interactionData.drag
-      } else {
-        return null
-      }
-    }, 'GhostOutline dragVector')
-
-    const colorTheme = useColorTheme()
-
-    if (frame == null || dragVector == null) {
+export function dragTargetsElementPaths(targets: Array<ElementPath>): DragTargetsElementPaths {
+  return {
+    type: 'element-paths',
+    targets: targets,
+  }
+}
+
+interface DragTargetsFrames {
+  type: 'frames'
+  frames: Array<CanvasRectangle>
+}
+
+export function dragTargetsFrame(frames: Array<CanvasRectangle>): DragTargetsFrames {
+  return {
+    type: 'frames',
+    frames: frames,
+  }
+}
+
+type DragOutlineControlProps = DragTargetsElementPaths | DragTargetsFrames
+
+export const DragOutlineControl = controlForStrategyMemoized((props: DragOutlineControlProps) => {
```

Comment:
probably nitpick, but it feels weird to me that `props` _is_ the tagged union type, with a `props.type` field on it.
Somehow it'd feel more natural to me if `props.targets` would be the union type. Again, nitpick, but it just sort of breaks my brain that this is the only props object in Utopia which is _also_ a tagged union.

Reply:
hmm but at the same time, if props.targets would be `DragTargetsElementPaths | DragTargetsFrames` then it'd bust the shallow equality. aaa! okay I changed my mind. it's good as it is :D

Reply:
Agree, and agree 😀